### DISCUSSION TITLE
UserInfo id null in onUserWillBeCreated callback

### DISCUSSION THREAD

**Discussion by Arnooodles** [serverpod/serverpod#2298] [Q&A] [answered]
I am currently integrating Google Sign-In with Serverpod and need to access the Google API before adding users. Therefore, I'm utilizing the `onUserWillBeCreated` callback. However, I'm encountering an issue where I cannot instantiate `GoogleAuth.authClientForUser(session, userId)` because the `userId` is null. Could this be a bug, or is there a step I've overlooked?

**Comment by Isakdl** [comment, marked as answer]
This is the intended behavior, the reason is that this callback is called before the user is created e.g. inserted into the database. Hence there is no id yet. This method can be used to stop a user from being created by returning false.

If you need the userId you have to use the callback `onUserCreated` that is called after the user has been inserted into the database.

**Reply by Arnooodles** [reply]
Thank you for clarifying this issue. I have one more question: what is the correct way to instantiate `GoogleAuth.authClientForUser` Should I use `session.auth.authenticatedUserId` or `userInfo.id`? According to the documentation on accessing the Google API, the `userId` is obtained from `authenticatedUserId`. However, I noticed that in the `onUserCreated` method, `authenticatedUserId` is null, while `userInfo.id` is not. When I tried to use the `userInfo.id`  it returns a null `googleClient`.

Can you confirm if it is possible to access the Google API within the onUserCreated method?

**Reply by Isakdl** [reply]
Both of these ids should be the same when using the serverpod_auth module. We should probably populate the id in the session so that it is available in the `onUserCreated` method but in any case using the userInfo.id is just as valid in this context.

**Reply by Arnooodles** [reply]
Could you please confirm whether the `googleClient` can be retrieved using the `GoogleAuth.authClientForUser` within the `onUserCreated` method? I am currently receiving a null value. Is there a step I might be overlooking? Below is the sample code for my current implementation for your review.
```
Future<void> _onUserCreated(Session session, UserInfo userInfo) async {
    try {
      final int? userId = userInfo.id;
      final AutoRefreshingAuthClient? googleClient = userId != null
          ? await GoogleAuth.authClientForUser(session, userId)
          : null;
      if (googleClient != null) {
        final PeopleServiceApi peopleService = PeopleServiceApi(googleClient);
        final Person person = await peopleService.people.get('people/me');
        ...
      }
      ...
 }
```

**Reply by Isakdl** [reply]
@Arnooodles I took a quick look at this. Very reasonable use case I think and it is supposed to work. However, it is not possible because of some implementation details in the framework but I considered that a bug so will push a fix for this.

**Reply by Arnooodles** [reply]
Thanks for looking into this! I appreciate you confirming the use case and identifying it as a bug. I'm glad to hear you'll be pushing a fix. Could you update me on which version I can use to test the fix for this issue?

**Reply by Isakdl** [reply]
The fix is merged into main and will be included in the next release!
